examples/kuadrant/reconcilers/istio.go

```diff
@@ -21,44 +21,52 @@ import (
 	"github.com/kuadrant/policy-machinery/machinery"
 )
 
-const istioGatewayProvider = "istio"
+const IstioGatewayProviderName = "istio"
 
 var (
-	_ GatewayProvider = &IstioGatewayProvider{}
-
-	istioAuthorizationPolicyKind       = schema.GroupKind{Group: istiov1.GroupName, Kind: "AuthorizationPolicy"}
-	istioAuthorizationPoliciesResource = istiov1.SchemeGroupVersion.WithResource("authorizationpolicies")
+	IstioAuthorizationPolicyKind       = schema.GroupKind{Group: istiov1.GroupName, Kind: "AuthorizationPolicy"}
+	IstioAuthorizationPoliciesResource = istiov1.SchemeGroupVersion.WithResource("authorizationpolicies")
 )
 
 type IstioGatewayProvider struct {
-	*dynamic.DynamicClient
+	Client *dynamic.DynamicClient
 }
 
-func (p *IstioGatewayProvider) ReconcileGateway(topology *machinery.Topology, gateway machinery.Targetable, capabilities map[string][][]machinery.Targetable) {
-	// check if the gateway is managed by the istio gateway controller
-	if !lo.ContainsBy(topology.Targetables().Parents(gateway), func(p machinery.Targetable) bool {
-		gc, ok := p.(*machinery.GatewayClass)
-		return ok && gc.Spec.ControllerName == "istio.io/gateway-controller"
-	}) {
-		return
-	}
-
-	// reconcile istio authorizationpolicy resources
-	paths := lo.Filter(capabilities["auth"], func(path []machinery.Targetable, _ int) bool {
-		return lo.Contains(lo.Map(path, machinery.MapTargetableToURLFunc), gateway.GetURL())
+func (p *IstioGatewayProvider) ReconcileAuthorizationPolicies(ctx context.Context, resourceEvent controller.ResourceEvent, topology *machinery.Topology) {
+	authPaths := pathsFromContext(ctx, authPathsKey)
+	targetables := topology.Targetables()
+	gateways := targetables.Items(func(o machinery.Object) bool {
+		_, ok := o.(*machinery.Gateway)
+		return ok
 	})
-	if len(paths) > 0 {
-		p.createAuthorizationPolicy(topology, gateway, paths)
-		return
+	for _, gateway := range gateways {
+		paths := lo.Filter(authPaths, func(path []machinery.Targetable, _ int) bool {
+			if len(path) < 4 { // should never happen
+				log.Fatalf("Unexpected topology path length to build Istio AuthorizationPolicy: %s\n", strings.Join(lo.Map(path, machinery.MapTargetableToURLFunc), " → "))
+			}
+			return path[0].GetURL() == gateway.GetURL() && lo.ContainsBy(targetables.Parents(path[0]), func(parent machinery.Targetable) bool {
+				gc, ok := parent.(*machinery.GatewayClass)
+				return ok && gc.Spec.ControllerName == "istio.io/gateway-controller"
+			})
+		})
+		if len(paths) > 0 {
+			p.createAuthorizationPolicy(topology, gateway, paths)
+			continue
+		}
+		p.deleteAuthorizationPolicy(topology, gateway.GetNamespace(), gateway.GetName(), gateway)
```

Comment:
```suggestion
		if len(paths) > 0 {
			p.createAuthorizationPolicy(ctx, topology, gateway, paths)
			continue
		}
		p.deleteAuthorizationPolicy(ctx, topology, gateway.GetNamespace(), gateway.GetName(), gateway)
```
Minor but since we have `ctx` passed in, we can use this context in the client create/update/delete calls in these methods and the same can be done in the equivalent in `envoy_gateway.go` if we change the method signature to take in a `Context`

Reply:
Done.